### ISSUE TITLE
Compare the result of nematus and marian-dev

### ISSUE THREAD

**Issue by 520jefferson** [marian-nmt/marian-dev#128]
I train CE with nematus and marian-dev,so confuse about the result ,the corpus is the same, and the configuration nearly the same.
1、nematus
       bleu       meter
C2E: 0.4112 | 0.616429
E2C: 0.5152 | 0.60306

2、marian-dev
         bleu    meter
C2E:  0.391 | 0.571157
E2C:  0.4564 | 0.536114

Configuration:
nematus:
train_len=100,
dim_word=512,  # word vector dimensionality
dim=1024,  # the number of LSTM units
factors=1, # input factors
dim_per_factor=None, # list of word vector dimensionalities (one per factor): [250,200,50] for total dimensionality of 500
encoder='gru',
decoder='gru_cond',
patience=10,  # early stopping patience
max_epochs=8000,
finish_after=20000000,  # finish after this many updates
dispFreq=10,
decay_c=0.,  # L2 regularization penalty
map_decay_c=0., # L2 regularization penalty towards original weights
clip_c=5.,  # gradient clipping threshold
lrate=0.001,  # learning rate
n_words_src=48550,  # source vocabulary size
n_words=31800,  # target vocabulary size
maxlen=50,  # maximum length of the description
optimizer='adam',
batch_size=256,
valid_batch_size=16,
saveto='models_1010/512-1024-ch_en',
validFreq=0,
saveFreq=2000,   # save the parameters after every saveFreq updates
sampleFreq=2000,   # generate some samples after every sampleFreq
datasets=[
    home + 'train0825.src.ch.0926.filter.bpe_integrate.shuf',
    home + 'train0825.tar.en.0926.filter.bpe_integrate.shuf'],
valid_datasets=['../data/dev/newstest2011.en.tok',
                '../data/dev/newstest2011.fr.tok'],
dictionaries=[
    home + 'train0825_CE/c2e_0926/train0825.src.ch.0926.filter.bpe_integrate.pkl',
    home + 'train0825_CE/c2e_0926/train0825.tar.en.0926.filter.bpe_integrate.pkl'],
use_dropout=True,
dropout_embedding=0.2, # dropout for input embeddings (0: no dropout)
dropout_hidden=0.2, # dropout for hidden layers (0: no dropout)
dropout_source=0, # dropout source words (0: no dropout)
dropout_target=0, # dropout target words (0: no dropout)
reload_=True,
reload_training_progress=True, # reload trainig progress (only used if reload_ is True)
overwrite=False,
external_validation_script=None,
shuffle_each_epoch=True,
sort_by_length=True,
use_domain_interpolation=False, # interpolate between an out-domain training corpus and an in-domain training corpus
domain_interpolation_min=0.1, # minimum (initial) fraction of in-domain training data
domain_interpolation_max=1.0, # maximum fraction of in-domain training data
domain_interpolation_inc=0.1, # interpolation increment to be applied each time patience runs out, until maximum amount of interpolation is reached
domain_interpolation_indomain_datasets=['indomain.en', 'indomain.fr'], # in-domain parallel training corpus
maxibatch_size=20, #How many minibatches to load at one time
objective="CE", #CE: cross-entropy; MRT: minimum risk training 
mrt_alpha=0.005,
mrt_samples=100,
mrt_samples_meanloss=10,
mrt_reference=True,
mrt_loss="SENTENCEBLEU n=4", # loss function for minimum risk training
mrt_ml_mix=0.5, # interpolate mrt loss with ML loss
model_version=0.1, #store version used for training for compatibility
prior_model=None, # Prior model file, used for MAP
tie_encoder_decoder_embeddings=False, # Tie the input embeddings of the encoder and the decoder (first factor only)
tie_decoder_embeddings=False, # Tie the input embeddings of the decoder with the softmax output embeddings
encoder_truncate_gradient=-1, # Truncate BPTT gradients in the encoder to this value. Use -1 for no truncation
decoder_truncate_gradient=-1, # Truncate BPTT gradients in the decoder to this value. Use -1 for no truncation

marian-dev:
../../build/marian \
--type amun \
--model models_amun/512-1024-ch_en.npz \
--devices $@ \
--seed 0 \
--dim-emb 512 \
--dim-rnn 1024 \
--train-set train0825.src.ch.0926.filter.bpe_integrate  train0825.tar.en.0926.filter.bpe_integrate \
--vocabs train0825.src.ch.0926.filter.bpe_integrate.pkl.json train0825.tar.en.0926.filter.bpe_integrate.pkl.json \
--dim-vocabs 48550 31800 \
--disp-freq  10 \
--save-freq  2000 \
--learn-rate 0.001 \
--max-length 50 \
--optimizer adam \
--mini-batch 256 \
--maxi-batch 20 \
--dropout-rnn 0.2 \
--dropout-src 0 \
--dropout-trg 0 \
--tempdir tmp \
--mini-batch-fit \
--workspace 15000 \
--clip-norm 5. \
--layer-normalization \
--after-batches 20000000

**Comment by emjotde** [comment]
Are you training with marian on multiple GPUs?

**Comment by 520jefferson** [comment]
Yes, both direction trained on 2 GPUS separately.

**Comment by emjotde** [comment]
These are large differences which I have never seen yet, comparing Nematus and Marian. 
I can only guess, but what might be happening is this: 

* You are using very large batches.
* You are using a very large learning rate.
* With multi-gpu training, the default mode is Asynchronous SGD. Async together with large batches will cause (more?) stale gradients, because it takes a while to update the common parameters. 
* Training get stuck because of the large learning rate and the stale gradients. 

To test this hypothesis it would be good to run the same experiment with the same settings with Marian on only a single GPU. Asynchronous SGD and different hyper-parameters may interact weirdly. 

* For how many iterations/epochs did you train both models?
* If you run another experiment with a single GPU, can you use `--log train.log` and send me the log to `junczys@amu.edu.pl`?

**Comment by 520jefferson** [comment]
First:
nematus:i change the code to adapt multi-gpus programs  and train enough,nearly  no bleu increase)
marin-dev:
c2e:
[2017-10-28 11:35:36] Ep. 13 : Up. 489040 : Sen. 13906152 : Cost 39.59 : Time 6.72s : 17039.30 words/s
[2017-10-28 11:35:43] Ep. 13 : Up. 489050 : Sen. 13913012 : Cost 31.82 : Time 6.42s : 17986.81 words/s
[2017-10-28 11:35:49] Ep. 13 : Up. 489060 : Sen. 13919302 : Cost 33.70 : Time 6.72s : 16647.73 words/s
e2c:(the bleu will drop at the back stage)
[2017-10-28 11:35:51] Ep. 12 : Up. 518010 : Sen. 18865315 : Cost 36.32 : Time 6.37s : 13088.12 words/s
[2017-10-28 11:35:58] Ep. 12 : Up. 518020 : Sen. 18870825 : Cost 40.52 : Time 6.29s : 16780.65 words/s

Second:Okay,i will train it on single GPU ,get the log  and try to send it to you.

**Comment by emjotde** [comment]
A general remark, it is usually helpful to have validation test set enabled during training to see how training behaves on unseen data, for instance whether you are overfitting or something. Maybe Marian reaches reasonable number earlier and then overfits or the model degrades somehow. Validation sets need to be preprocessed in the same way as the training data. 

Both, Marian and Nematus use them.

**Comment by emjotde** [comment]
Just for clarification, you said: "nematus:i change the code to adapt multi-gpus programs and train enough,nearly no bleu increase)"

Do you mean you are training with Nematus on multiple GPUs?

**Comment by 520jefferson** [comment]
Yes, i put both c2e and e2c on the GPU 0 1 2 3.

**Comment by emjotde** [comment]
So, would that be a Synchronous SGD implementation in **Nematus** on 4 GPUs, right?

**Comment by 520jefferson** [comment]
We use Easgd in Nematus to update params.

**Comment by emjotde** [comment]
OK, that's important information and seems to point towards my stale gradients hypothesis. I have no good intuition how EASGD and large learning rates interact, but that makes the training a lot less comparable. If we want to completely rule out interactions like that I would repeat the training with Marian but set:

```
--learn-reate 0.0003
--lr-warmup 16000
--lr-decay-inv-sqrt 16000
--lr-report
--sync-sgd
--devices 0 1 2 3
```
This will enable `synchronous` SGD which is as stable as single GPU training and has no stale gradients. It will split batches across GPUs, so batches will be up to 4x larger than normal, but usually that helps with convergence. I would lower the learning rate to `0.0003`, futher more that setting introduces a learning-rate warmup phase during the first 16000 iterations and then will slowly decrease the learning rate from 0.0003 starting at iteration 16000. `--lr-report` displays the current learning rate at the end of a log line.

**Comment by 520jefferson** [comment]
I already start up the  E2C  training process with  the configures you said above. 
[2017-10-28 17:09:59] Ep. 1 : Up. 10 : Sen. 17120 : Cost 181.33 : Time 85.00s : 3517.85 words/s : L.r. 1.8750e-07
[2017-10-28 17:10:16] Ep. 1 : Up. 20 : Sen. 34520 : Cost 210.84 : Time 16.08s : 21772.34 words/s : L.r. 3.7500e-07
[2017-10-28 17:10:30] Ep. 1 : Up. 30 : Sen. 51200 : Cost 183.61 : Time 14.41s : 20456.55 words/s : L.r. 5.6250e-07

**Comment by emjotde** [comment]
Looks good. Now let's see how it goes.

**Comment by 520jefferson** [comment]
The epochs i have trained as follows,the top4 bleu below .
iter | bleu | meter
72000 | 0.4604 | 0.541099
94000 | 0.4604 | 0.542596
96000 | 0.4599 | 0.542217
92000 | 0.4586 | 0.539705

[2017-10-30 10:14:11] Ep. 8 : Up. 101220 : Sen. 10158080 : Cost 34.04 : Time 13.99s : 21002.52 words/s : L.r. 1.1927e-04
[2017-10-30 10:14:26] Ep. 8 : Up. 101230 : Sen. 10175120 : Cost 35.22 : Time 14.31s : 21823.36 words/s : L.r. 1.1927e-04
[2017-10-30 10:14:40] Ep. 8 : Up. 101240 : Sen. 10192240 : Cost 33.70 : Time 14.09s : 21374.18 words/s : L.r. 1.1926e-04

**Comment by emjotde** [comment]
So, there is still a 5 BLEU points difference?

**Comment by 520jefferson** [comment]
Yes, so confused about the result. anything else i can do to improve the bleu.

**Comment by emjotde** [comment]
I am as confused as you. Can you share the models and the test sets only? That would be a couple of hundred MB. Preferably for both, Marian and Nematus.

**Comment by 520jefferson** [comment]
We check the codes, as to marian-dev, we find that when testing the model  we ignore  the bpe ( old nematus has itself) before we get the bleu. 
So i will use bpe in testing. If the effect is not so good i will figure out to get the model and test sets out for you .

**Comment by 520jefferson** [comment]
Sorry,It's my mistake too, when i use bpe before translate then the score is up. Then i will test all of the models.
iter    bleu    meter
108000 0.5280 0.5937461475876906

By the way,i didn't find codes about bpe ,whether it will be added later?

**Comment by emjotde** [comment]
* Currently there are not plans to add BPE to marian, but that might change. It's not a high priority.
* warmp and decay are not randomized. Seed is a initialization value for random events in Marian, like corpus shuffling or parameter initialization. So there should be no relation between the two other than resulting from shuffling and initialization. 

In that case I am closing this issue.
